File: 2025/day_01/day1.py
import os.path
import unittest
import time

def line_split(line: str) :
    return (line[0], int(line[1:]))

def parse_input(file_path: str) :
    if not os.path.exists(file_path) :
        print(f"File {file_path} not found.")
        return None
    with open(file_path, 'r') as f :
        return list(map(line_split, f.read().split('\n')))



def r1(puzzle_input, debug=False) :
    if puzzle_input is None:
        return None
    res = 0
    current = 50
    for x in puzzle_input :
        direction, value = x
        if (direction == 'L'):
            current -= value
        if (direction == 'R'):
            current += value
        current %= 100
        if current == 0:
            res += 1
    print(f"Final position: {current}")
    return res



def r2(puzzle_input, debug=False) :
    if puzzle_input is None:
        return None
    res = 0
    current = 50
    for x in puzzle_input :
        direction, value = x
        if (direction == 'L'):
            while value > 0:
                current -= 1
                value -= 1
                current %= 100
                if current == 0:
                    res += 1
        if (direction == 'R'):
            while value > 0:
                current += 1
                value -= 1
                current %= 100
                if current == 0:
                    res += 1
    print(f"Final position: {current}")
    return res



class TestsOfToday(unittest.TestCase):

    def setUp(self):
        pass

    def test(self):
        self.assertEqual(1, 1)

if __name__ == '__main__':
    unittest.main(exit=False)

    example = parse_input('input-example.txt')
    print("Example input:")
    print(example)

    puzzle = parse_input('input.txt')

    print("--- Part One ---")
    t0 = time.time()
    print(f"Example result: {r1(example, True)}")
    print(f"Puzzle answer:  {r1(puzzle)}")
    t1 = time.time()
    print(f"Part one computed in {t1 - t0} seconds.")

    print("--- Part Two ---")
    t0 = time.time()
    print(f"Example result: {r2(example, True)}")
    print(f"Puzzle answer:  {r2(puzzle)}")
    t1 = time.time()
    print(f"Part two computed in {t1 - t0} seconds.")

File: 2025/day_01/Day1.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.Duration;
import java.util.List;

public class Day1 {

    // written with jdk 25
    // javac Day1.java; java Day1
    public static void main(String[] args) {

        var example = parseInput("input-example.txt");
        System.out.println("Example input:");
        printInput(example);

        var puzzle = parseInput("input.txt");

        System.out.println("--- Part One ---");
        Instant start1 = Instant.now();
        System.out.println("Example answer: " + part1(example));
        System.out.println("Puzzle answer:  " + part1(puzzle));
        Instant finish1 = Instant.now();
        System.out.println("Part one computed in " + formatDuration(start1, finish1));

        System.out.println("--- Part Two ---");
        Instant start2 = Instant.now();
        System.out.println("Example answer: " + part2(example));
        System.out.println("Puzzle answer:  " + part2(puzzle));
        Instant finish2 = Instant.now();
        System.out.println("Part two computed in " + formatDuration(start2, finish2));
    }

    static String formatDuration(Instant instantA, Instant instantB) {
        var duration = Duration.between(instantA, instantB);
        long msValue = duration.toMillis();
        return msValue < 1000 ? msValue + " ms" : duration.toSeconds() + " seconds";
    }

    // Input parsing

    enum Direction {
        RIGHT, LEFT;

        static Direction of(String value) {
            return switch (value) {
                case "L" -> LEFT;
                case "R" -> RIGHT;
                default -> throw new IllegalArgumentException(value);
            };
        }
    }

    record Rotation(Direction direction, Integer value) {}

    static List<Rotation> parseInput(String fileName) {
        try {
            var parsed = Files.readAllLines(Path.of(fileName)).stream()
                .map(line -> new Rotation(
                    Direction.of(line.substring(0, 1)), 
                    Integer.parseInt(line.substring(1))))
                .toList();
            System.out.println(fileName + " successfully parsed.");
            return parsed;
        } catch (IOException ioException) {
            System.out.println(ioException.getMessage());
            return null;
        }
    }

    static void printInput(List<Rotation> puzzleInput) {
        puzzleInput.forEach(rotation -> System.out.println(rotation));
    }

    // Part One

    static Integer part1(List<Rotation> rotations) {
        if (rotations == null)
            return null;
        int result = 0;
        int current = 50;
        for (Rotation rotation : rotations) {
            switch (rotation.direction()) {
                case LEFT -> current -= rotation.value();
                case RIGHT -> current += rotation.value();
            }
            current %= 100;
            if (current == 0)
                result ++;
        }
        return result;
    }

    // Part Two

    static Integer part2(List<Rotation> rotations) {
        if (rotations == null)
            return null;
        Clock clock = new Clock();
        return rotations.stream()
            .map(rotation -> clock.rotate(rotation.direction(), rotation.value()))
            .reduce(0, Integer::sum);
    }

    static class Clock {
        private int current = 50;

        public int rotate(Direction direction, int value) {
            int ticks = 0;
            while (value > 0) {
                switch (direction) {
                    case LEFT -> this.current -= 1;
                    case RIGHT -> this.current += 1;
                }
                current %= 100;
                if (current == 0)
                    ticks ++;
                value --;
            }
            return ticks;
        }
    }

}
